presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.util.maps;
+
+import com.google.common.base.Equivalence;
+import org.intellij.lang.annotations.Flow;
+
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * A Map decorator allowing for overriding the equivalence relation
+ * for the underlying Map's keys.
+ * <p>
+ * EquivalenceMap instances retain iteration order of the underlying Map.
+ * Collections returned from #entrySet() and #keySet() are unmodifiable
+ * views.
+ * <p>
+ * Similarly to java.util.IdentityHashMap this class violates parts of
+ * Map's contract. See java.util.IdentityHashMap's javadoc for details.
+ *
+ * @param <K>
+ * @param <V>
+ */
+public class EquivalenceMap<K, V>
+        implements Map<K, V>
+{
+    private final Map<Equivalence.Wrapper<K>, V> delegate;
+    private final Equivalence<Object> equivalence;
+
+    public static <K, V> EquivalenceMap<K, V> on(MapSupplier supplier, Equivalence<Object> equivalence)
```

Comment:
This can't be a `Supplier<Map<K,V>>`, because the underlying Map is `Map<Equivalence.Wrapper<K>, V>`. Notice that MapSupplier's only method is generic on return type but the MapSupplier itself is not. This has the desired consequences descibed in my comment here: https://github.com/Teradata/presto/pull/413#discussion_r87391666